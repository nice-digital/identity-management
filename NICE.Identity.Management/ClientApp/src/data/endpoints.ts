const API_BASE_URL = process.env.REACT_APP_API_BASE_URL
	? process.env.REACT_APP_API_BASE_URL
	: "#{REACT_APP_API_BASE_URL}";

export const Endpoints = {
	usersList: `${API_BASE_URL}/users`, // get all users
	user: (userId: string | number): string => `${API_BASE_URL}/users/${userId}`,

	servicesList: `${API_BASE_URL}/services`,
	service: (serviceId: string | number): string =>
		`${API_BASE_URL}/services/${serviceId}`,
	
	userRolesByWebsite: (userId: string | number, websiteId: string | number): string =>
		`${API_BASE_URL}/users/${userId}/rolesbywebsite/${websiteId}`,

	verificationEmail: `${API_BASE_URL}/VerificationEmail/VerificationEmail`,
	websitesList: `${API_BASE_URL}/websites`,
};
