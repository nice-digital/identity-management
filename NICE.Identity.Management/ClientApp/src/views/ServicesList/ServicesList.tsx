import React, { Component } from "react";

import { Link } from "react-router-dom";
import { Breadcrumbs, Breadcrumb } from "@nice-digital/nds-breadcrumbs";
import { Card } from "@nice-digital/nds-card";
import { Grid, GridItem } from "@nice-digital/nds-grid";
import { PageHeader } from "@nice-digital/nds-page-header";
import {
	queryStringToObject,
	appendQueryParameter,
	removeQueryParameter,
	stripMultipleQueries,
} from "src/utils/querystring";

import { fetchData } from "src/helpers/fetchData";
import { isDataError } from "src/helpers/isDataError";
import { Endpoints } from "src/data/endpoints";
import { type HistoryType, type WebsiteType } from "src/models/types";
import { FilterSearch } from "src/components/FilterSearch/FilterSearch";
import { FilterBox } from "src/components/FilterBox/FilterBox";
import { WebsiteEnvironment } from "src/components/WebsiteEnvironment/WebsiteEnvironment";
import { ErrorMessage } from "src/components/ErrorMessage/ErrorMessage";
import { Pagination } from "src/components/Pagination/Pagination";

import styles from "./ServicesList.module.scss";

type CardMetaData = {
	label?: string;
	value: React.ReactNode;
};

type ServicesListProps = {
	basename: string;
	location: {
		pathname: string;
		search: string;
	};
	history: HistoryType;
};

type ServicesListState = {
	path: string;
	originalWebsites: Array<WebsiteType>;
	websites: Array<WebsiteType>;
	searchQuery?: string;
	error?: Error;
	isLoading: boolean;
	environmentFiltersChecked: Array<string>;
	pageNumber: number;
	itemsPerPage: number | string;
	environmentsForFilter: Array<string>;
};

export class ServicesList extends Component<
	ServicesListProps,
	ServicesListState
> {
	constructor(props: ServicesListProps) {
		super(props);

		const querystring = this.props.location.search;

		const querystringObject = queryStringToObject(querystring);

		const pageNumber = Number(
			querystringObject.page
				? Array.isArray(querystringObject.page)
					? querystringObject.page[0]
					: querystringObject.page
				: 1,
		);

		let itemsPerPage = querystringObject.amount
			? Array.isArray(querystringObject.amount)
				? querystringObject.amount[0]
				: querystringObject.amount
			: 25;

		itemsPerPage = Number(itemsPerPage) ? Number(itemsPerPage) : itemsPerPage;

		this.state = {
			path: "",
			originalWebsites: [],
			websites: [],
			isLoading: true,
			pageNumber: pageNumber,
			itemsPerPage: itemsPerPage,
			environmentFiltersChecked: [],
			environmentsForFilter: [],
		};

		document.title = "NICE Accounts - Services list";
	}

	async componentDidMount(): Promise<void> {
		this.setState({ isLoading: true });

		const websites = await fetchData(Endpoints.websitesList);

		if (isDataError(websites)) {
			this.setState({ error: websites });
		} else {
			const allEnvironments = websites.map(
				(website: { environment: { name: string } }) =>
					website.environment.name,
			);

			const environmentsForFilter = allEnvironments.reduce(function (
				accumulatedEnvironments: string[],
				currentEnvironment: string,
			) {
				if (accumulatedEnvironments.indexOf(currentEnvironment) === -1) {
					accumulatedEnvironments.push(currentEnvironment);
				}
				return accumulatedEnvironments;
			},
			[]);

			this.setState({ environmentsForFilter });
		}

		this.setState({ originalWebsites: websites, websites, isLoading: false });
	}

	pastPageRange = (
		itemsPerPage: string | number,
		pageNumber: number,
		dataCount: number,
	): number => {
		let pastPageRange = false;

		if (Number(itemsPerPage)) {
			itemsPerPage = Number(itemsPerPage);
			pastPageRange = pageNumber * itemsPerPage >= dataCount;
		}

		if (pastPageRange || itemsPerPage === "all") {
			pageNumber = 1;
		}

		return pageNumber;
	};

	filterWebsitesByEnvironment = (environment: string): void => {
		this.setState({ isLoading: true });

		let environmentFiltersChecked = this.state.environmentFiltersChecked;

		environmentFiltersChecked = environmentFiltersChecked.includes(environment)
			? environmentFiltersChecked.filter(
					(environmentFilter) => environmentFilter !== environment,
					)
			: environmentFiltersChecked.concat(environment);

		const itemsPerPage = Number(this.state.itemsPerPage)
			? Number(this.state.itemsPerPage)
			: this.state.itemsPerPage;

		let websites = this.state.originalWebsites,
			pageNumber = this.state.pageNumber;

		if (environmentFiltersChecked.length)
			websites = this.websitesByEnvironment(
				environmentFiltersChecked,
				websites,
			);

		pageNumber = this.pastPageRange(
			itemsPerPage,
			pageNumber,
			this.state.websites.length,
		);

		this.setState({
			websites,
			environmentFiltersChecked,
			pageNumber,
			isLoading: false,
		});
	};

	filterWebsitesBySearch = async (searchQuery: string): Promise<void> => {
		this.setState({ isLoading: true });

		const originalWebsites = await fetchData(
			`${Endpoints.websitesList}?q=${searchQuery}`,
		);

		let websites = originalWebsites,
			pageNumber = this.state.pageNumber;

		const itemsPerPage = Number(this.state.itemsPerPage)
			? Number(this.state.itemsPerPage)
			: this.state.itemsPerPage;

		if (isDataError(originalWebsites)) {
			this.setState({ error: originalWebsites });
		}

		if (this.state.environmentFiltersChecked.length) {
			websites = this.websitesByEnvironment(
				this.state.environmentFiltersChecked,
				websites,
			);
		}

		pageNumber = this.pastPageRange(
			itemsPerPage,
			pageNumber,
			this.state.websites.length,
		);

		this.setState({
			originalWebsites,
			websites,
			searchQuery,
			pageNumber,
			isLoading: false,
		});
	};

	websitesByEnvironment = (
		environmentFiltersChecked: Array<string>,
		websites: Array<WebsiteType>,
	): Array<WebsiteType> => {
		return (websites = websites.filter((website) => {
			const websiteEnvironment = website.environment.name;

			if (environmentFiltersChecked.includes(websiteEnvironment)) {
				return website;
			}
		}));
	};

	getPaginateStartAndFinishPosition = (
		websitesCount: number,
		pageNumber: number,
		itemsPerPage: number | string,
	): { start: number; finish: number } => {
		const paginationPositions = {
			start: 0,
			finish: websitesCount,
		};

		const itemAmountIsNumber = Number(itemsPerPage);

		if (itemAmountIsNumber) {
			paginationPositions.start = (pageNumber - 1) * itemAmountIsNumber;
			paginationPositions.finish =
				paginationPositions.start + itemAmountIsNumber <= websitesCount
					? paginationPositions.start + itemAmountIsNumber
					: websitesCount;
		}

		return paginationPositions;
	};

	getPaginationText = (
		websitesCount: number,
		start: number,
		finish: number,
	): string => {
		const amountPerPage = finish - start;
		const paginationExtract =
			websitesCount > amountPerPage ? `${start + 1} to ${finish} of ` : "";

		return `Showing ${paginationExtract}${websitesCount} service${
			websitesCount === 1 ? "" : "s"
		}`;
	};

	changeAmount = (e: React.ChangeEvent<HTMLSelectElement>): void => {
		let itemsPerPage = (e.target as HTMLSelectElement).value || 25,
			pageNumber = this.state.pageNumber,
			path = stripMultipleQueries(this.state.path, ["amount", "page"]);

		itemsPerPage = Number(itemsPerPage) ? Number(itemsPerPage) : itemsPerPage;

		pageNumber = this.pastPageRange(
			itemsPerPage,
			pageNumber,
			this.state.websites.length,
		);

		path = appendQueryParameter(path, "amount", itemsPerPage.toString());
		path = appendQueryParameter(path, "page", pageNumber);

		this.setState({ itemsPerPage, path, pageNumber }, () => {
			this.props.history.push(path);
		});
	};

	changePage = (e: React.MouseEvent<HTMLAnchorElement>): void => {
		e.preventDefault();

		let pageNumber =
				(e.target as HTMLAnchorElement).getAttribute("data-pager") || 1,
			path = removeQueryParameter(this.state.path, "page");

		if (pageNumber === "previous") {
			pageNumber = this.state.pageNumber - 1;
		}
		if (pageNumber === "next") {
			pageNumber = this.state.pageNumber + 1;
		}

		pageNumber = Number(pageNumber);
		path = appendQueryParameter(path, "page", pageNumber);

		this.setState({ pageNumber, path }, () => {
			this.props.history.push(path);
		});
	};

	render(): JSX.Element {
		const {
			websites,
			searchQuery,
			error,
			isLoading,
			pageNumber,
			itemsPerPage,
			environmentFiltersChecked,
		} = this.state;

		const paginationPositions = this.getPaginateStartAndFinishPosition(
			websites.length,
			pageNumber,
			itemsPerPage,
		);

		const paginationText = this.getPaginationText(
			websites.length,
			paginationPositions.start,
			paginationPositions.finish,
		);

		const websitesPaginated = websites.length
			? websites.slice(paginationPositions.start, paginationPositions.finish)
			: websites;

		return (
			<>
				<Breadcrumbs>
					<Breadcrumb to="/overview" elementType={Link}>
						Administration
					</Breadcrumb>
					<Breadcrumb>Services</Breadcrumb>
				</Breadcrumbs>

				{!error ? (
					<>
						<PageHeader heading="Services" />
						<Grid>
							<GridItem cols={12} md={3}>
								<FilterSearch
									onInputChange={this.filterWebsitesBySearch}
									label={"Filter by service name or URL"}
								/>

								<FilterBox
									name="Environments"
									filters={this.state.environmentsForFilter}
									selected={environmentFiltersChecked}
									onCheckboxChange={this.filterWebsitesByEnvironment}
									hideFilterPanelHeading={true}
								/>
							</GridItem>
							<GridItem cols={12} md={9} aria-busy={!websites.length}>
								{isLoading ? (
									<p>Loading...</p>
								) : websites.length ? (
									<>
										<h2
											className={styles.servicesListSummary}
											data-qa-sel="services-returned"
										>
											{paginationText}
										</h2>
										<ul className="list--unstyled" aria-label="websites" data-qa-sel="list-of-websites">
											{websitesPaginated.map((website) => {
												const { id, host, service } = website;
												const servicesListHeading = {
													headingText: `${service.name}`,
													link: {
														elementType: Link,
														destination: `/websites/${id}`,
													},
												};

												const servicesListMetadata: Array<CardMetaData> = [
													{
														value: <WebsiteEnvironment website={website} />,
													},
													{
														label: "Website",
														value: host,
													},
												];

												return (
													<li key={id}>
														<Card
															{...servicesListHeading}
															metadata={servicesListMetadata}
															key={id}
														/>
													</li>
												);
											})}
										</ul>
										<Pagination
											onChangePage={this.changePage}
											onChangeAmount={this.changeAmount}
											itemsPerPage={itemsPerPage}
											consultationCount={websites.length}
											currentPage={pageNumber}
										/>
									</>
								) : searchQuery ? (
									<p>No results found for &quot;{searchQuery}&quot;</p>
								) : (
									<p>No results found</p>
								)}
							</GridItem>
						</Grid>
					</>
				) : (
					<>
						<PageHeader heading="Error" />
						<ErrorMessage error={error}></ErrorMessage>
					</>
				)}
			</>
		);
	}
}
