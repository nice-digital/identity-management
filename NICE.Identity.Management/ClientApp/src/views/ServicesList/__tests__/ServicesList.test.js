import React from "react";
import { mount, shallow } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import toJson from "enzyme-to-json";

import { ServicesList } from "../ServicesList";
import websites from "./websites.json";

import { nextTick } from "../../../utils/nextTick";

import { Endpoints } from "../../../data/endpoints";

describe("ServicesList", () => {
	const servicesListProps = {
		location: {
			pathname: "/services",
			search: "?amount=all&page=1",
		},
		history: {
			push: jest.fn(),
		},
	};

	const servicesListPropsOnePerPage = {
		...servicesListProps,
		location: { search: "?amount=1&page=4" },
	};

	const servicesListPropsThreePerPage = {
		...servicesListProps,
		location: { search: "?amount=3&page=1" },
	};
	
	const dummyText = "SomeText";

	const consoleErrorReset = console.error;

	beforeEach(() => {
		fetch.resetMocks();
		console.error = consoleErrorReset;
	});

	it("should show loading message before data has been loaded", () => {
		fetch.mockResponseOnce(JSON.stringify(websites));
		const wrapper = shallow(<ServicesList {...servicesListProps} />);
		expect(wrapper.find("p").text()).toEqual("Loading...");
	});

	// it("should call fetchData during componentDidMount", () => {
	// 	fetch.mockResponseOnce(JSON.stringify(websites));
	// 	const wrapper = mount(<MemoryRouter><ServicesList {...servicesListProps} /></MemoryRouter>);
	// 	const spy = jest.spyOn(wrapper.instance(), "componentDidMount");
	// 	wrapper.instance().componentDidMount();
	// 	wrapper.update();
	// 	expect(spy).toHaveBeenCalled();
	// 	expect(fetch.mock.calls.length).toEqual(1);
	// 	expect(fetch.mock.calls[0][0]).toEqual(Endpoints.websites);
	// 	spy.mockClear();
	// });

	it("should match the snapshot after data has been loaded", async () => {
		fetch.mockResponseOnce(JSON.stringify(websites));
		const wrapper = mount(
			<MemoryRouter>
				<ServicesList {...servicesListProps} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		expect(toJson(wrapper, { noKey: true, mode: "deep" })).toMatchSnapshot();
	});

	it("should show error message when fetch returns 401 error", async () => {
		console.error = jest.fn();		
		fetch.mockResponseOnce(JSON.stringify({}), { status: 401 });
		const wrapper = mount(<MemoryRouter><ServicesList {...servicesListProps} /></MemoryRouter>);
		await nextTick();
		wrapper.update();
		expect(toJson(wrapper, { noKey: true, mode: "deep" })).toMatchSnapshot();
	});

	it("should show error message when fetch returns 500 error", async () => {
		console.error = jest.fn();
		fetch.mockRejectOnce(new Error("500 Internal Server Error"));
		const wrapper = mount(<MemoryRouter><ServicesList {...servicesListProps} /></MemoryRouter>);
		await nextTick();
		wrapper.update();
		expect(toJson(wrapper, { noKey: true, mode: "deep" })).toMatchSnapshot();
	});

	it("should show no results message when fetch returns an empty array", async () => {
		fetch.mockResponseOnce(JSON.stringify([]));
		const wrapper = shallow(<ServicesList {...servicesListProps} />);
		await nextTick();
		wrapper.update();
		expect(wrapper.find("p").text()).toEqual("No results found");
	});

	// it("should show no results found message after search returns empty array", async () => {
	// 	fetch.mockResponseOnce(JSON.stringify(websites));
	// 	fetch.mockResponseOnce(JSON.stringify([]));
	// 	const wrapper = shallow(<ServicesList {...servicesListProps} />);
	// 	const instance = wrapper.instance();
	// 	await nextTick();
	// 	wrapper.update();
	// 	instance.filterUsersBySearch(dummyText);
	// 	await nextTick();
	// 	wrapper.update();
	// 	expect(wrapper.find("p").text()).toEqual(
	// 		`No results found for ${dummyText}`,
	// 	);
	// });

	// it("should show all filter by default", () => {
	// 	fetch.mockResponseOnce(JSON.stringify(websites));
	// 	const wrapper = mount(<MemoryRouter><ServicesList {...usersListProps} /></MemoryRouter>);
	// 	expect(wrapper.find("#filter-status-all").props().defaultChecked).toEqual(
	// 		true,
	// 	);
	// });

	// it("should filter users to all active when radio button is clicked", async () => {
	// 	fetch.mockResponseOnce(JSON.stringify(websites));
	// 	const wrapper = mount(
	// 		<MemoryRouter>
	// 			<ServicesList {...usersListProps} />
	// 		</MemoryRouter>,
	// 	);
	// 	await nextTick();
	// 	wrapper.update();
	// 	wrapper.find("#filter-status-active").simulate("change", {
	// 		target: { value: "active" },
	// 	});
	// 	await nextTick();
	// 	wrapper.update();
	// 	wrapper.find(".tag").forEach((tag) => {
	// 		expect(tag.text()).toEqual("Active");
	// 	});
	// });

	// it("should show 25 (default page amount) or less results by default when paginated", async () => {
	// 	fetch.mockResponseOnce(JSON.stringify(websites));
	// 	const wrapper = mount(
	// 		<MemoryRouter>
	// 			<ServicesList {...usersListProps} />
	// 		</MemoryRouter>,
	// 	);
	// 	await nextTick();
	// 	wrapper.update();
	// 	const listContainer = wrapper.find("[data-qa-sel='list-of-users']");
	// 	expect(listContainer.find(".card").length).toBeLessThanOrEqual(25);
	// });

	// it("should go to page 2 when next button is clicked", async () => {
	// 	fetch.mockResponseOnce(JSON.stringify(websites));
	// 	const wrapper = mount(
	// 		<MemoryRouter>
	// 			<ServicesList {...usersListPropsThreePerPage} />
	// 		</MemoryRouter>,
	// 	);
	// 	await nextTick();
	// 	wrapper.update();
	// 	wrapper.find("[data-pager='next']").simulate("click");
	// 	await nextTick();
	// 	wrapper.update();
	// 	const listContainer = wrapper.find("[data-qa-sel='list-of-users']");
	// 	const usersListSummary = wrapper.find(".usersListSummary");
	// 	expect(usersListSummary.text()).toEqual("Showing 4 to 4 of 4 users");
	// 	expect(wrapper.find(".paginationCounter").text()).toEqual("Page 2 of 2");
	// 	expect(listContainer.find(".card").length).toEqual(1);
	// });

	// it("should go to first page when page 1 button is clicked", async () => {
	// 	fetch.mockResponseOnce(JSON.stringify(websites));
	// 	const wrapper = mount(
	// 		<MemoryRouter>
	// 			<ServicesList {...usersListPropsOnePerPage} />
	// 		</MemoryRouter>,
	// 	);
	// 	await nextTick();
	// 	wrapper.update();
	// 	wrapper.find("[data-pager='1']").simulate("click");
	// 	await nextTick();
	// 	wrapper.update();
	// 	const listContainer = wrapper.find("[data-qa-sel='list-of-users']");
	// 	const usersListSummary = wrapper.find(".usersListSummary");
	// 	expect(usersListSummary.text()).toEqual("Showing 1 to 1 of 4 users");
	// 	expect(wrapper.find(".paginationCounter").text()).toEqual("Page 1 of 4");
	// 	expect(listContainer.find(".card").length).toEqual(1);
	// });
});
