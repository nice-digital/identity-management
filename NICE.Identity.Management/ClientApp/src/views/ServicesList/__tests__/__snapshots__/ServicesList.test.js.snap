// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServicesList should match the snapshot after data has been loaded 1`] = `
Array [
  Array [
    <nav
      aria-label="Breadcrumbs"
      role="navigation"
    >
      <ol
        className="breadcrumbs"
      >
        <li
          className="breadcrumbs__crumb"
        >
          <a
            href="/overview"
            onClick={[Function]}
          >
            Administration
          </a>
        </li>
        <li
          className="breadcrumbs__crumb"
        >
          <span>
            Services
          </span>
        </li>
      </ol>
    </nav>,
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "{
  \\"@context\\": \\"http://schema.org\\",
  \\"@type\\": \\"BreadcrumbList\\",
  \\"itemListElement\\": [
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 1,
      \\"item\\": {
        \\"@id\\": \\"/overview\\",
        \\"name\\": \\"Administration\\"
      }
    },
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 2,
      \\"item\\": {
        \\"name\\": \\"Services\\"
      }
    }
  ]
}",
        }
      }
      type="application/ld+json"
    />,
  ],
  <div
    className="page-header"
  >
    <h1
      className="page-header__heading"
    >
      Services
    </h1>
  </div>,
  <div
    className="grid"
  >
    <div
      data-g="12 md:3"
    >
      <div
        className="panel"
      >
        <div
          className="input"
        >
          <label
            className="input__label"
            htmlFor="user-search"
          >
            Filter by service name or URL
          </label>
          <input
            autoComplete="off"
            className="input__input"
            id="user-search"
            name="user-search"
            onChange={[Function]}
            type="search"
            unique="userSearch"
          />
        </div>
      </div>
      <div
        className="filter-panel-hide-heading"
      >
        <div
          className="filter-panel undefined"
        >
          <h2
            className="filter-panel__heading"
          >
            <a
              aria-controls="filter-panel-body"
              aria-expanded="true"
              href="#filter-panel-body"
            />
          </h2>
          <div
            aria-hidden={false}
            className="filter-panel__body"
            id="filter-panel-body"
          >
            <div
              className="filter-group"
            >
              <h3
                className="filter-group__heading"
              >
                <button
                  aria-controls="group-environments"
                  aria-expanded={true}
                  onClick={[Function]}
                  type="button"
                >
                  <span
                    id="group-heading-environments"
                  >
                    Environments
                  </span>
                </button>
              </h3>
              <fieldset
                aria-hidden={false}
                className="filter-group__options"
                id="group-environments"
              >
                <legend>
                  Environments
                </legend>
                <label
                  className="filter-option"
                  htmlFor="filter_environments_alpha"
                >
                  <input
                    checked={false}
                    id="filter_environments_alpha"
                    name="environments"
                    onChange={[Function]}
                    title="Environments - Alpha"
                    type="checkbox"
                    value="Alpha"
                  />
                  <span
                    className="filter-option__text"
                  >
                    Alpha
                  </span>
                </label>
                <label
                  className="filter-option"
                  htmlFor="filter_environments_test"
                >
                  <input
                    checked={false}
                    id="filter_environments_test"
                    name="environments"
                    onChange={[Function]}
                    title="Environments - Test"
                    type="checkbox"
                    value="Test"
                  />
                  <span
                    className="filter-option__text"
                  >
                    Test
                  </span>
                </label>
                <label
                  className="filter-option"
                  htmlFor="filter_environments_dev"
                >
                  <input
                    checked={false}
                    id="filter_environments_dev"
                    name="environments"
                    onChange={[Function]}
                    title="Environments - Dev"
                    type="checkbox"
                    value="Dev"
                  />
                  <span
                    className="filter-option__text"
                  >
                    Dev
                  </span>
                </label>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      aria-busy={false}
      data-g="12 md:9"
    >
      <h2
        className="servicesListSummary"
      >
        Showing 5 services
      </h2>
      <ul
        className="list--unstyled"
        data-qa-sel="list-of-websites"
      >
        <li>
          <article
            className="card"
          >
            Array [
              <header
                className="card__header"
              >
                <p
                  className="card__heading"
                >
                  Comment Collection
                </p>
              </header>,
              <dl
                className="card__metadata"
              >
                <div
                  className="card__metadatum"
                >
                  <dd>
                    <span
                      className="tag tag--alpha"
                      data-qa-sel="environment-alpha"
                    >
                      Alpha
                    </span>
                  </dd>
                </div>
                <div
                  className="card__metadatum"
                >
                  <dt
                    className="visually-hidden"
                  >
                    Website
                  </dt>
                  <dd>
                    alpha-comments.nice.org.uk
                  </dd>
                </div>
              </dl>,
            ]
          </article>
        </li>
        <li>
          <article
            className="card"
          >
            Array [
              <header
                className="card__header"
              >
                <p
                  className="card__heading"
                >
                  EPPI Reviewer
                </p>
              </header>,
              <dl
                className="card__metadata"
              >
                <div
                  className="card__metadatum"
                >
                  <dd>
                    <span
                      className="tag tag--new"
                      data-qa-sel="environment-test"
                    >
                      Test
                    </span>
                  </dd>
                </div>
                <div
                  className="card__metadatum"
                >
                  <dt
                    className="visually-hidden"
                  >
                    Website
                  </dt>
                  <dd>
                    test-eppi.nice.org.uk
                  </dd>
                </div>
              </dl>,
            ]
          </article>
        </li>
        <li>
          <article
            className="card"
          >
            Array [
              <header
                className="card__header"
              >
                <p
                  className="card__heading"
                >
                  Idam Docker
                </p>
              </header>,
              <dl
                className="card__metadata"
              >
                <div
                  className="card__metadatum"
                >
                  <dd>
                    <span
                      className="tag tag--new"
                      data-qa-sel="environment-test"
                    >
                      Test
                    </span>
                  </dd>
                </div>
                <div
                  className="card__metadatum"
                >
                  <dt
                    className="visually-hidden"
                  >
                    Website
                  </dt>
                  <dd>
                    test-idam
                  </dd>
                </div>
              </dl>,
            ]
          </article>
        </li>
        <li>
          <article
            className="card"
          >
            Array [
              <header
                className="card__header"
              >
                <p
                  className="card__heading"
                >
                  Comment Collection
                </p>
              </header>,
              <dl
                className="card__metadata"
              >
                <div
                  className="card__metadatum"
                >
                  <dd>
                    <span
                      className="tag tag--updated"
                      data-qa-sel="environment-dev"
                    >
                      Dev
                    </span>
                  </dd>
                </div>
                <div
                  className="card__metadatum"
                >
                  <dt
                    className="visually-hidden"
                  >
                    Website
                  </dt>
                  <dd>
                    dev-comments.nice.org.uk
                  </dd>
                </div>
              </dl>,
            ]
          </article>
        </li>
        <li>
          <article
            className="card"
          >
            Array [
              <header
                className="card__header"
              >
                <p
                  className="card__heading"
                >
                  EPPI Reviewer
                </p>
              </header>,
              <dl
                className="card__metadata"
              >
                <div
                  className="card__metadatum"
                >
                  <dd>
                    <span
                      className="tag tag--updated"
                      data-qa-sel="environment-dev"
                    >
                      Dev
                    </span>
                  </dd>
                </div>
                <div
                  className="card__metadatum"
                >
                  <dt
                    className="visually-hidden"
                  >
                    Website
                  </dt>
                  <dd>
                    dev-eppi.nice.org.uk
                  </dd>
                </div>
              </dl>,
            ]
          </article>
        </li>
      </ul>
      <div
        className="flex flexAlignCenter"
      >
        <div
          className="mr--e"
          data-qa-sel="number-of-results-on-page"
        >
          <label
            className="bold mr--c"
            htmlFor="itemsPerPage"
          >
            Items per page
          </label>
          <select
            data-qa-sel="result-on-the-page-index"
            id="itemsPerPage"
            name="itemsPerPage"
            onChange={[Function]}
            value="all"
          >
            <option
              value="25"
            >
              25
            </option>
            <option
              value="50"
            >
              50
            </option>
            <option
              value="all"
            >
              All
            </option>
          </select>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`ServicesList should show error message when fetch returns 401 error 1`] = `
Array [
  Array [
    <nav
      aria-label="Breadcrumbs"
      role="navigation"
    >
      <ol
        className="breadcrumbs"
      >
        <li
          className="breadcrumbs__crumb"
        >
          <a
            href="/overview"
            onClick={[Function]}
          >
            Administration
          </a>
        </li>
        <li
          className="breadcrumbs__crumb"
        >
          <span>
            Services
          </span>
        </li>
      </ol>
    </nav>,
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "{
  \\"@context\\": \\"http://schema.org\\",
  \\"@type\\": \\"BreadcrumbList\\",
  \\"itemListElement\\": [
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 1,
      \\"item\\": {
        \\"@id\\": \\"/overview\\",
        \\"name\\": \\"Administration\\"
      }
    },
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 2,
      \\"item\\": {
        \\"name\\": \\"Services\\"
      }
    }
  ]
}",
        }
      }
      type="application/ld+json"
    />,
  ],
  <div
    className="page-header"
  >
    <h1
      className="page-header__heading"
    >
      Services
    </h1>
  </div>,
  <div
    className="alert alert--error"
  >
    <h2>
      Whoops
    </h2>
    <p>
      Sorry, there's been an error!
    </p>
  </div>,
]
`;

exports[`ServicesList should show error message when fetch returns 500 error 1`] = `
Array [
  Array [
    <nav
      aria-label="Breadcrumbs"
      role="navigation"
    >
      <ol
        className="breadcrumbs"
      >
        <li
          className="breadcrumbs__crumb"
        >
          <a
            href="/overview"
            onClick={[Function]}
          >
            Administration
          </a>
        </li>
        <li
          className="breadcrumbs__crumb"
        >
          <span>
            Services
          </span>
        </li>
      </ol>
    </nav>,
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "{
  \\"@context\\": \\"http://schema.org\\",
  \\"@type\\": \\"BreadcrumbList\\",
  \\"itemListElement\\": [
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 1,
      \\"item\\": {
        \\"@id\\": \\"/overview\\",
        \\"name\\": \\"Administration\\"
      }
    },
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 2,
      \\"item\\": {
        \\"name\\": \\"Services\\"
      }
    }
  ]
}",
        }
      }
      type="application/ld+json"
    />,
  ],
  <div
    className="page-header"
  >
    <h1
      className="page-header__heading"
    >
      Services
    </h1>
  </div>,
  <div
    className="alert alert--error"
  >
    <h2>
      Whoops
    </h2>
    <p>
      Sorry, there's been an error!
    </p>
  </div>,
]
`;
